addOns/client/src/main/java/org/zaproxy/addon/client/internal/ClientMapWriter.java

```diff
@@ -0,0 +1,181 @@
+        fw.write(output);
+        return false;
+    }
+
+    private static void outputNode(Writer fw, ClientNode node, int level) throws IOException {
+        if (node.isStorage()) {
+            // Skip storage nodes in the tree
+            // Those details are represented as components of their source
+            return;
+        }
+        // We could create a set of data structures and use jackson, but the format is very
+        // simple - it still uses jackson for value output
+        String indent = " ".repeat(level * 2);
+
+        outputKV(
+                fw,
+                indent,
+                true,
+                NODE_KEY,
+                level == 0 ? ROOT_NODE_NAME : node.getUserObject().getName());
+
+        if (node.getUserObject().isStorage()) {
+            outputKV(fw, indent, false, STORAGE_KEY, node.getUserObject().isStorage());
+        }
+        if (!node.getUserObject().isVisited()) {
+            outputKV(fw, indent, false, VISITED_KEY, node.getUserObject().isVisited());
+        }
+        if (node.getUserObject().getComponents() != null
+                && !node.getUserObject().getComponents().isEmpty()) {
+            for (ClientSideComponent component : node.getUserObject().getComponents()) {
+                if (component.getType() == ClientSideComponent.Type.REDIRECT) {
+                    outputKV(
+                            fw, indent, false, component.getType().getLabel(), component.getHref());
+                } else if (component.getType() == ClientSideComponent.Type.CONTENT_LOADED) {
+                    outputKV(fw, indent, false, component.getType().getLabel(), true);
+                }
+            }
+            indent = outputComponents(fw, node.getUserObject().getComponents(), level, indent);
+        }
+
+        Stats.incCounter(ExtensionClientIntegration.PREFIX + ".export.clientmap.node");
+
+        if (node.getChildCount() > 0) {
+            fw.write(indent);
+            fw.write("  ");
+            fw.write(CHILDREN_KEY);
+            fw.write(":");
+            fw.write('\n');
+            node.children()
+                    .asIterator()
+                    .forEachRemaining(
+                            c -> {
+                                try {
+                                    outputNode(fw, (ClientNode) c, level + 1);
+                                } catch (IOException e) {
+                                    throw new UncheckedIOException(e);
+                                }
+                            });
+        }
+    }
+
+    private static String outputComponents(
+            Writer fw, Set<ClientSideComponent> components, int level, String indent)
+            throws IOException {
+        fw.write(indent);
+        fw.write("  ");
+        fw.write(COMPONENTS_KEY);
+        fw.write(":\n");
+
+        indent = " ".repeat((level + 1) * 2);
+        SortedSet<ClientSideComponent> sortedComponents = new TreeSet<>(components);
```

Comment:
This can be reverted to what was before (though fine as is too), although less like to have concurrency issues while exporting it would be better to sync on it to prevent them.

Reply:
Was the other more efficient somehow?

Reply:
This seems more concise to me 🤷‍♂️

Reply:
Changed to:
```java
        SortedSet<ClientSideComponent> sortedComponents =
                Collections.synchronizedSortedSet(new TreeSet<>(components));
```

Sufficient?

Reply:
Should sync on `components`.